kratos/utilities/function_parser_utility.h

```diff
@@ -0,0 +1,261 @@
+    ///@{
+
+    /// The index type definition
+    typedef std::size_t IndexType;
+
+    /// Counted pointer of GenericFunctionUtility
+    KRATOS_CLASS_POINTER_DEFINITION(GenericFunctionUtility);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /**
+     * @brief Default constructor
+     * @param rFunctionBody The string defining the function
+     * @param LocalSystem The parameters defining the local system
+     */
+    GenericFunctionUtility(
+        const std::string& rFunctionBody,
+        Parameters LocalSystem = Parameters{}
+        );
+
+    ///Copy constructor
+    GenericFunctionUtility(GenericFunctionUtility const& rOther);
+
+    /// Destructor.
+    ~GenericFunctionUtility();
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief This method clones the current function instance
+     * @return A clone of the current class
+     */
+    Pointer Clone()
+    {
+        return Kratos::make_shared<GenericFunctionUtility>(*this);
+    }
+
+    /**
+     * @brief This method returns if it depends on space
+     * @return True if it uses the local system, false otherwise
+     */
+    bool UseLocalSystem()
+    {
+        return mUseLocalSystem;
+    }
+
+    /**
+     * @brief This method returns if it depends on space
+     * @return True if it depends on space, false otherwise
+     */
+    bool DependsOnSpace()
+    {
+        return mDependsOnSpace;
+    }
+
+    /**
+     * @brief This method returns the function body
+     * @return The function body
+     */
+    std::string FunctionBody();
+
+    /**
+     * @brief This method rotates and calls the evaluation function
+     * @param x The x coordinate
+     * @param y The y coordinate
+     * @param z The z coordinate
+     * @param t The time variable
+     * @param X The initial x coordinate
+     * @param Y The initial y coordinate
+     * @param Z The initial z coordinate
+     */
+    double RotateAndCallFunction(
+        const double x,
+        const double y,
+        const double z,
+        const double t,
+        const double X = 0.0,
+        const double Y = 0.0,
+        const double Z = 0.0
+        );
+
+    /**
+     * @brief This calls the evaluation function
+     * @param x The x coordinate
+     * @param y The y coordinate
+     * @param z The z coordinate
+     * @param t The time variable
+     * @param X The initial x coordinate
+     * @param Y The initial y coordinate
+     * @param Z The initial z coordinate
+     */
+    double CallFunction(
+        const double x,
+        const double y,
+        const double z,
+        const double t,
+        const double X = 0.0,
+        const double Y = 0.0,
+        const double Z = 0.0
+        );
+
+    ///@}
+
+private:
+
+    ///@name Member Variables
+    ///@{
+
+    std::unordered_map<std::string, double>  mNameSpace;   /// The variables considered on the function
+    te_expr* mpTinyExpr = nullptr;                         /// The function parser
```

Comment:
are u explicitly handling the delete? (i cannot see where if you do)

maybe it is best if u use a unique_ptr to handle this.

Reply:
Yes, see the destructor

Reply:
but then, why don't you use an unique_ptr? then the deletion would happen automatically

Reply:
`unique_ptr` does not work with forward declaration

and the forward declaration is required to not leak the implementation of the tinyexpr into our headers

Reply:
i am a bit lost here.

are u sure you cannot define

~~~cpp
Kratos::unique_ptr<te_expr> mpTinyExpr = nullptr?
~~~

i honestly cannot see why that should not work

Reply:
https://stackoverflow.com/questions/6012157/is-stdunique-ptrt-required-to-know-the-full-definition-of-t

Seems like it might work with `unique_ptr`

but the times I tried I didn't get it to work and used `shared_ptr` in the end

Reply:
The API provides a destructor, I am using te_free as the documentation indicates:

~~~c
    #include "tinyexpr.h"
    #include <stdio.h>

    int main(int argc, char *argv[])
    {
        if (argc < 2) {
            printf("Usage: example2 \"expression\"\n");
            return 0;
        }

        const char *expression = argv[1];
        printf("Evaluating:\n\t%s\n", expression);

        /* This shows an example where the variables
         * x and y are bound at eval-time. */
        double x, y;
        te_variable vars[] = {{"x", &x}, {"y", &y}};

        /* This will compile the expression and check for errors. */
        int err;
        te_expr *n = te_compile(expression, vars, 2, &err);

        if (n) {
            /* The variables can be changed here, and eval can be called as many
             * times as you like. This is fairly efficient because the parsing has
             * already been done. */
            x = 3; y = 4;
            const double r = te_eval(n); printf("Result:\n\t%f\n", r);
            te_free(n);
        } else {
            /* Show the user where the error is at. */
            printf("\t%*s^\nError near here", err-1, "");
        }

        return 0;
    }
~~~